xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -246,31 +250,44 @@ public void shutdown() {
         "methodConfig", Collections.singletonList(methodConfig.build()));
   }
 
-  @VisibleForTesting
-  static Map<String, ?> generateServiceConfigWithLoadBalancingConfig(Collection<String> clusters) {
+  // called in syncContext
+  private Map<String, ?> generateServiceConfigWithLoadBalancingConfig() {
     Map<String, Object> childPolicy = new HashMap<>();
-    for (String cluster : clusters) {
-      List<Map<String, Map<String, String>>> lbPolicy =
-          Collections.singletonList(
-              Collections.singletonMap(
-                  "cds_experimental", Collections.singletonMap("cluster", cluster)));
-      childPolicy.put(cluster, Collections.singletonMap("lbPolicy", lbPolicy));
-    }
-    return Collections.singletonMap("loadBalancingConfig",
-        Collections.singletonList(
-            Collections.singletonMap(
-                "cluster_manager_experimental", Collections.singletonMap(
-                    "childPolicy", Collections.unmodifiableMap(childPolicy)))));
+    for (String name : clusterRefs.keySet()) {
+      String traditionalCluster = clusterRefs.get(name).traditionalCluster;
+      if (traditionalCluster != null) {
+        List<?> lbPolicy = ImmutableList.of(
+            ImmutableMap.of("cds_experimental", ImmutableMap.of("cluster", traditionalCluster)));
+        childPolicy.put(name, ImmutableMap.of("lbPolicy", lbPolicy));
+      }
+      RlsPluginConfig rlsPluginConfig = clusterRefs.get(name).rlsPluginConfig;
+      if (rlsPluginConfig != null) {
+        Map<String, ?> rlsConfig = new ImmutableMap.Builder<String, Object>()
+            .put("routeLookupConfig", rlsPluginConfig.config())
+            .put(
+                "childPolicy",
+                ImmutableList.of(ImmutableMap.of("cds_experimental", ImmutableMap.of())))
+            .put("childPolicyConfigTargetFieldName", "cluster")
+            .build();
+        Map<String, ?> rlsPolicy = ImmutableMap.of("rls_experimental", rlsConfig);
+        childPolicy.put(name, ImmutableMap.of("lbPolicy", ImmutableList.of(rlsPolicy)));
+      }
```

Comment:
This is totally a nit. I was thinking how to improve readability and flow of this a bit more, and came up with this:

```suggestion
      Map<String, ?> policy;
      ClusterRefState clusterRef = clusterRefs.get(name);
      if (clusterRef.traditionalCluster != null) {
        policy = ImmutableMap.of(
            "cds_experimental",
            ImmutableMap.of("cluster", clusterRef.traditionalCluster));
      } else if (clusterRef.rlsPluginConfig != null) {
        policy = clusterRef.rlsPluginConfig.toPolicy();
      } else {
        throw new IllegalStateException("ClusterRefState unexpected state.");
      }
      childPolicy.put(name, ImmutableMap.of("lbPolicy", ImmutableList.of(policy)));
```

This and add this to `RlsPluginConfig`:

```java
    public Map<String, ?> toPolicy() {
      return ImmutableMap.of("rls_experimental", new ImmutableMap.Builder<String, Object>()
          .put("routeLookupConfig", config())
          .put("childPolicy",
              ImmutableList.of(ImmutableMap.of("cds_experimental", ImmutableMap.of())))
          .put("childPolicyConfigTargetFieldName", "cluster")
          .build());
    }
```

To goal is to get to the common denominator - the output of both plugin types is a map. So we reduce some duplication around adding a child policy (`childPolicy.put()`).

It might be even better if `RlsPluginConfig` and the traditional cluster had a common interface with `toPolicy` method, f.e. `ClusterSpecifier`. This will make it simpler to deal with them in general.

Reply:
Great suggestion. I leave common interface with `ClusterSpecifier` unchanged because I can not predict whether all the plugins are associated with lb policy.

Reply:
I'll add `toLbPolicy()` method for `ClusterRefState`.

Reply:
Thank you!